packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,256 @@
+                    <span>Do Something</span>
+                </nimble-button>
+        `;
+    }
+}
+
+Angular template:
+
+<nimble-table>
+    <nimble-table-column-button #tableButton>Press Me</nimble-table-column-button>
+</nimble-table>
+
+Angular code:
+
+@ViewChild('tableButton', { read: NimbleTableColumnButton, static: true }) private readonly tableButton!: NimbleTableColumnButton;
+
+public ngOnInit(): void {
+    tableButton.callback = this.doSomething;
+}
+
+private doSomething(id: string): void {
+    ...
+}
+
+```
+
+Note the missing implementation in the above `ITableColumn` implementations are the necessary pieces to register them as FAST components.
+
+### Initial Nimble-provided Columns
+
+`TableColumnText`
+
+_Custom Properties:_
+
+```TS
+placeholder: string; // The string to display while no value is present
+```
+
+`TableColumnHyperlink`
+
+The API for this column type is out of scope for this spec and should be defined in a separate document. A specific concern for this column that must be addressed is how the hyperlinks that are placed in the shadow DOM of the table will work with the Angular Router (for instance).
+
+`TableColumnIcon`
+
+The API for this column type is also out of scope and should be covered in a separate document. Some unique concerns for this column is how a user can specify the set of icons to display, and then how a user can specify the logic to associate a particular value with one of the provided icons. Additionally, how do we allow a user to specify custom header content (won't be a static `headerTemplateFn` `ITableColumn` implementation).
+
+## Alternative Implementations / Designs
+
+### Programmatic API
+
+A programmatic API was also considered either in place of, or along side the proposed declarative API. Since declaring what columns to show is an aspect of view configuration, it makes sense to accomplish this declaratively when possible. Offering a programmatic API alongside the declarative one, while possible, does introduce complexity in the implementation that would be nice to avoid, at least initially, if possible.
+
+## Open Issues
+
+-   The current design doesn't offer any strict templating feedback for a particular `ITableColumn` implementation. So, if a user provides a dataKey to a property of an `ITableColumn` that wants the value for that dataKey to be a `DateTime` (i.e. its `cellTemplateFn` implementation expects a `DateTime`), but the value in the actual table data for that key is a string, the user will be unaware of that mismatch at compile time.
+
+    It is unclear how we could provide such feedback, but it would be extremely nice if possible.
+
+-   There is some concern with the lifecycle of these custom column components.
+-   Should `ITableCellData` also require a data index? This could enable some scenarios that might otherwise require a user to have id fields in their data.
```

Comment:
My thought is that we shouldn't be doing anything special with the data index (assuming that is the index of a row within the `data` property on the table). I think for a lot of usages (at least in SLE), that value will be almost meaningless. Instead, I could see the table having a concept of a row ID that is specified with a data key provided to the table. If none is specified, the row index is used. Then, that value of id/index could be passed to the cell data, if we liked that idea.

Reply:
Removed. If we decide we need such a member to be defined on the `TableCellData` interface we can introduce it then.